janusgraph-server/src/main/java/org/janusgraph/graphdb/tinkerpop/gremlin/server/auth/HMACAuthenticator.java

```diff
@@ -0,0 +1,222 @@
+            hmacAlgo = config.get(CONFIG_HMAC_ALGO).toString();
+        } else {
+            hmacAlgo = DEFAULT_HMAC_ALGO;
+        }
+
+        if (null != config && config.containsKey(CONFIG_TOKEN_TIMEOUT)) {
+            timeout = ((Number) config.get(CONFIG_TOKEN_TIMEOUT)).longValue();
+        } else {
+            timeout = DEFAULT_HMAC_TOKEN_TIMEOUT;
+        }
+
+        super.setup(config);
+
+        if (null != config & config.containsKey(CONFIG_HMAC_SECRET)) {
+            secret = config.get(CONFIG_HMAC_SECRET).toString().toCharArray();
+        } else {
+            secret = DEFAULT_HMAC_SECRET;
+        }
+    }
+
+    @Override
+    public AuthenticatedUser authenticate(final Map<String, String> credentials) throws AuthenticationException {
+        if (credentials.get(PROPERTY_GENERATE_TOKEN) != null) {
+            final AuthenticatedUser user = authenticateUser(credentials);
+            if (user == null) {
+                throw new AuthenticationException(AUTH_ERROR);
+            }
+            credentials.put(PROPERTY_TOKEN, getToken(credentials));
+            return user;
+        } else if (credentials.get(PROPERTY_TOKEN) != null) {
+            if (validateToken(credentials)) {
+                return new AuthenticatedUser(credentials.get(PROPERTY_USERNAME));
+            } else {
+                throw new AuthenticationException("Invalid token");
+            }
+        } else {
+            return authenticateUser(credentials);
+        }
+    }
+
+    private AuthenticatedUser authenticateUser(final Map<String, String> credentials) throws AuthenticationException {
+        final Vertex v = credentialStore.findUser(credentials.get(PROPERTY_USERNAME));
+        if (null == v || !BCrypt.checkpw(credentials.get(PROPERTY_PASSWORD), v.value(PROPERTY_PASSWORD))) {
+            throw new AuthenticationException(AUTH_ERROR);
+        }
+        return new AuthenticatedUser(credentials.get(PROPERTY_USERNAME));
+    }
+
+    private boolean validateToken(Map<String, String> credentials) {
+        final String token = credentials.get(PROPERTY_TOKEN);
+        final Map<String, String> tokenMap = parseToken(token);
+        final String username = tokenMap.get(PROPERTY_USERNAME);
+        final String time = tokenMap.get("time");
+        final String password = credentialStore.findUser(username).value(PROPERTY_PASSWORD);
+        final String salt = getBcryptSaltFromStoredPassword(password);
+        final String expected = generateToken(username, salt, time);
+        final Long timeLong = Long.parseLong(time);
+        final Long currentTime = new Date().getTime();
+        final String base64Token = new String(Base64.getUrlEncoder().encode(token.getBytes()));
+
+        if (timeLong + timeout < currentTime) {
+            return false;
+        } else {
+            //Don't short circuit comparison to prevent timing attacks
+            boolean isValid = true;
+            for (int i = 0; i < token.length(); i++) {
```

Comment:
Here the HMAC is rendered useless because it is short circuited when the incoming length truncates it. 
You need to check first that `expected.length() == base64Token.length()`
I would drive the loop by `expected.length()`.
Either way, it would seem that you should use either `token` or `base64Token` but not both.